### PR TITLE
Stubs for torch.nn

### LINKED ISSUE #18724: Type annotations for 'nn' module missing
## 🚀 Feature

These can be mostly autogenerated. 
cf https://github.com/pytorch/pytorch/issues/16996.

### PR DESCRIPTION
**Pull request by malmaud**
Closes https://github.com/pytorch/pytorch/issues/18724

### CODE REVIEW

**Comment by malmaud** [comment]
Don't merge yet, work in progress

**Comment by malmaud** [comment]
After diving into this, I think automating the stub creation for the modules (`nn.Module` and descendants) is not practical. There are ultimately too many special cases and idiosyncratic methods and types in play. The relationship between the type signature of the function a module wraps and the type signature of the various methods of the module and the module properties isn't easily expressed with a simple algorithm.

I think I landed on a reasonable compromise. I autogenerated stubs for `nn.functional` and used mypy's `stubgen` program to create candidate stubs for the modules, which I then manually edited to have more precise stubs. 

I am still doing some codegen on the module stubs to automatically create a `__call__` method for each module that defines a `forward` method and giving `__call__` the same type signature as `forward`. Ideally there would be some way to express in the type system that `__call__` and `forward` have the same type signature for all subtypes of `Module`, but that is AFAIK not possible.

There is still some cleanup to do, but this is getting close.

**Comment by malmaud** [comment]
I don't think I'll have time for a while to make these stubs much better than they are now in the current version of this PR. My plan now is to get this PR to pass CI and then step away.

The current status of things is every public method in the `nn` module has a stub, and the stubs for all the provided `nn.Module` subclasses are tight. 

The type annotations for some other methods, especially those in `nn.functional`, are overly permissive (ie, all correct code will be reported as correct but some type-incorrect code with not raise errors in `mypy`). That's no worse than the current status quo without type stubs, so I believe there's no harm in including them.  Anyone who's motivated could come around later and submit a PR to make them tighter.

How does that sound, @ezyang?

**Comment by malmaud** [comment]
Actually, I think even the loose stubs for `nn.functional` are an improvement on the status quo since at least PyCharm/VSCode will now be able to autocomplete all the methods in them, even if still can't infer the name and types of the arguments of all those methods.

**Comment by ezyang** [comment]
Incomplete/more permissive stubs sounds fine to me. Do you want me to review now, or wait till you resolve merge conflicts?

**Comment by malmaud** [comment]
Might as well wait; I'll ping you when it's ready for review. Thanks!

**Comment by shtratos** [comment]
@malmaud thank you for the titanic effort! Do you need any help with completing this?

**Comment by ezyang** [comment]
@malmaud previously stated that he wouldn't have time to push this all
through to completion; if you'd like to take over that would be much
appreciated :)

Excerpts from Dmitry Stratiychuk's message of 2019-06-24 15:54:35 -0700:
> @malmaud thank you for the titanic effort! Do you need any help with completing this?
>

**Comment by malmaud** [comment]
Thanks @shtratos ! Maybe the easiest path is I get this in a mergeable state this week, and then you can submit a follow-up PR to tighten up any of the stubs that too loose. The only thing left to do on this PR really is tidy it up and test that the stubs are correctly installed.

**Comment by malmaud** [comment]
CI passes now!

**Comment by ezyang** [comment]
So I'm guessing, not WIP anymore :)

File: tools/pyi/gen_pyi.py [hunk 1]
```diff
@@ -313,6 +314,82 @@ def generate_type_hints(fname, decls, is_tensor=False):
 
     return type_hints
 
+def gen_nn_modules(out):
+    def replace_forward(m):
+        # We instruct mypy to not emit errors for the `forward` and `__call__` declarations since mypy
+        # would otherwise correctly point out that Module's descendants' `forward` declarations 
+        # conflict with `Module`s. Specificlaly, `Module` defines `forward(self, *args)` while the 
+        # descandantes define more specific forms, such as `forward(self, input: Tensor)`, which
```
Comment (ezyang) [inline tools/pyi/gen_pyi.py]: nit: descendents

File: tools/pyi/gen_pyi.py [hunk 2]
```diff
@@ -313,6 +314,82 @@ def generate_type_hints(fname, decls, is_tensor=False):
 
     return type_hints
 
+def gen_nn_modules(out):
+    def replace_forward(m):
+        # We instruct mypy to not emit errors for the `forward` and `__call__` declarations since mypy
+        # would otherwise correctly point out that Module's descendants' `forward` declarations 
+        # conflict with `Module`s. Specificlaly, `Module` defines `forward(self, *args)` while the
```
Comment (ezyang) [inline tools/pyi/gen_pyi.py]: nit: Specifically

File: torch/CMakeLists.txt [hunk 3]
```diff
@@ -248,23 +248,108 @@ if (USE_NCCL)
     endif()
 endif()
 
-add_custom_target(torch_python_stubs DEPENDS "${TORCH_SRC_DIR}/__init__.pyi")
+# In the most recent CMake versions, a new 'TRANSFORM' subcommand of 'list' allows much of the boilerplate of defining the lists
+# of type stub files to be omitted. 
+# For comptability with older CMake versions, we omit it for now, but leave it as a comment in case comptability with the older
```
Comment (ezyang) [inline torch/CMakeLists.txt]: nit: compatibility

**Review by ezyang — APPROVED** [review]
Impressive.

**Comment by ezyang** [comment]
I did not do a careful review of the annotations. Let us treat these as beta for now and let people submit bugfixes for errors as things go.

**Pull request closed without merging**